Add function calling

rpeden opened the pull request:
This PR adds function calling to the Kestra OpenAI plugin. With these changes included, it's possible to use OpenAI function completions in a Kestra flow like so:

```yaml
id: hello-ai
namespace: awesome-project
tasks:
  - id: cool-ai-function
    type: io.kestra.plugin.openai.ChatCompletion
    model:  gpt-3.5-turbo
    apiKey: <api-key>
    messages: 
      - role: user
        content: My cat really hated this food! He's normally not picky, but after 
                 one bite of this food he actually picked up the entire bag of 
                 food and dropped it in the trash!
    functions: 
      - name: classify_review
        description: Classify a product review entered on our website as positive, neutral, 
                     or negative and generate responses to the review.
        parameters: 
          - name: classification
            description: The rating of the product review's positivity
            type: string
            required: true
            enumValues:  
              - negative
              - neutral
              - positive
          - name: friendly_reply
            description: A thoughtful from the merchant to the customer suitable for posting on an e-commerce website.
            type: string
            required: true
          - name: hostile_reply
            description: A scathing reply the merchant should send if they never want to hear from this customer again.
            type: string
            required: true
    functionCall: classify_review
```

and get function call output from OpenAI:

![Screenshot 2023-12-23 213501](https://github.com/kestra-io/plugin-openai/assets/228762/7da46c33-317d-4064-ac46-969a94af1d61)

rpeden committed:
Added function calling to `ChatCompletion` plugin

rpeden committed:
Updated AbstractTask to customize OpenAiService creation

`com.theokanning.openai` function calling implementation assumes you want to call a Java method based on OpenAI's response; it doesn't give you a way to extract the JSON response of the function call unless you add a Jackson mixin.

rpeden committed:
Added function calling to plugin

rpeden committed:
Added function calling unit test

rpeden commented on src/main/java/io/kestra/plugin/openai/AbstractTask.java:
```diff
@@ -20,9 +31,34 @@ public abstract class AbstractTask extends Task implements OpenAiInterface{
+        abstract Class<?> getParametersClass();
+
+        @JsonProperty("parameters")
+        abstract ArrayNode getParameters();
```
`com.theokanning.openai` assumes that you want to use the result of OpenAI function calls to call Java methods; it doesn't give you the option of getting the JSON response from OpenAI unless you customize the creation of the `OpenAiApi` class. This required the creation both a mixin class and a class that extends the `com.theokanning.openai` library's `ChatFunction`.

loicmathieu replied:
This is unfortunate, maybe this issue should be brought to the OpenAI client so that a more elegant solution would be possible.
Also, can you explain in a comment why the Mixin is needed (what you describe in this comment).

rpeden replied:
Fortunately, the OpenAI client library ended up merging a PR that made it possible to call functions without this workaround.

loicmathieu commented on src/main/java/io/kestra/plugin/openai/AbstractTask.java:
```diff
@@ -20,9 +31,34 @@ public abstract class AbstractTask extends Task implements OpenAiInterface{
+        title = "The timeout value to send to the OpenAI API."
+    )
+    @PluginProperty
+    private Integer apiTimeout;
```
We previously didn't have this timeout is it really needed?

rpeden replied:
Sorry, I meant to comment on this too. 

OkHttp's default timeout of 10 seconds is too short for many OpenAI API calls, especially function calls. 

But perhaps it's better to set the timeout on the OkHttp client to a large value and then just use the Task's `timeout` value so the flow can move on if the OpenAI call doesn't finish quickly enough.

loicmathieu replied:
There is two ways to fix this issue:
- hardcode a larger value (for ex 30s), this is simpler but may be limited
- do as you did but with a default value that match the default of OkHttp (so 10s).

loicmathieu commented on src/main/java/io/kestra/plugin/openai/ChatCompletion.java:
```diff
@@ -67,6 +64,19 @@ public class ChatCompletion extends AbstractTask implements RunnableTask<ChatCom
+    @Schema(
+        title = "The function call(s) the API can use when generating completions."
+    )
+    @PluginProperty
```
This should be dynamic and it's content rendered as done with ChatMessage

rpeden replied:
I updated this to (I think) render everything that needs to be rendered.

loicmathieu commented on src/main/java/io/kestra/plugin/openai/ChatCompletion.java:
```diff
@@ -210,6 +232,32 @@ public static class Output implements io.kestra.core.models.tasks.Output {
     }
 
+    @Getter
+    public static class PluginChatFunction extends ChatFunction {
```
Is this class needed? Can't you do the same as with ChatMessage?

rpeden replied:
Unfortunately, `ChatMessage` only lets you set the `parameters` type as a `Class` - it's implemented as 

```
@JsonProperty("parameters")
    private Class<?> parametersClass;
```

`parameters` is a JSON schema, so the way `ChatMessage` is written there's no way to pass through the values from the Kestra flow at runtime. from This ties in the the mixin above. Using `PluginChatFunction` alone doesn't work, because adds its own mixin here: https://github.com/TheoKanning/openai-java/blob/442cf130f6fbedbdcb70ce9d96ce572faf54c9e9/service/src/main/java/com/theokanning/openai/service/OpenAiService.java#L423

The only way I was able to override it was creating `PluginChatFunction` to provide the ability to set the `parameters` JSON schema at runtime, and then add a new mixin to prevent Jackson from falling back on the underlying `ChatFunctionCallMixIn`. 

This is all far uglier than I'd like. When I looked at the library's code a few weeks ago, it looked like he'd designed `ChatFunction` to work the way it does, and offered the ability to override things the way I did for anyone who needed different behavior. But perhaps it is worth asking him. :)

rpeden replied:
I changed this so it no longer extends ChatFunction. The library adds a class called `ChatFunctionDynamic` that enables function calling without needing the workarounds I used before, but when I try to use 'ChatFunctionDynamic` directly, Kestra says: 

```
Cannot construct instance of `com.theokanning.openai.completion.chat.ChatFunctionDynamic` (although at least one Creator exists): cannot deserialize from Object value (no delegate- or property-based Creator)
```

Creating classes for functions and parameters solved this and also made it possible to add `Schema` annotations to provide helpful guidance in the Kestra editor.

loicmathieu reviewed (changes requested):
Thanks a lot!

rpeden committed:
Merge branch 'master' into add-function-calling

rpeden committed:
Updated to latest release of `com.theokanning.openai-gpt3-java`

rpeden committed:
Revert changes to `AbstractTask`

rpeden committed:
Updated `ChatCompletion` to use new library functionality and incorporate earlier feedback

rpeden committed:
Added function calling example

rpeden committed:
Updating existing function call test and added a test that checks enum value function calls

rpeden committed:
Merge branch 'master' into add-function-calling

rpeden committed:
Update test to latest gpt-3.5-turbo model

rpeden commented:
@anna-geller Since you assigned this to me, I figured I'd better hurry up and finish it! 😅 

The OpenAI wrapper library this plugin uses _finally_ merged a PR that had been sitting for months. The PR fixes the issue I did an ugly workaround for, so I reverted the ugliness, updated the code to use the new library functionality, and added another test. Hopefully it's okay, but I'm happy to address any remaining issues. I'm not 100% sure I did the `render` calls correctly.

anna-geller commented:
Awesome 👍 sorry for the assignment, this was not me but my bot that automatically assigned PR authors as assignees

Thanks so much for contributing, especially over the holidays!

Skraye committed:
fix(): cleanup

Skraye commented:
I did the cleanup for the rendering and made a small fix to make sure all tests passed. Thanks for your contribution!

rpeden commented:
@Skraye thank you! I apologize for not cleaning up the imports - I thought I had done that, but clearly I missed it.
